### PR TITLE
Support Alibaba-NLP/gte-Qwen2-7B-instruct embedding Model

### PR DESCRIPTION
**Pull request by zhaochenyang20**
## Motivation

Current SGLang only supports the e5-mistral embedding model. I added Alibaba-NLP/gte-Qwen2-7B-instruct model in this PR. 

Also, previously SGLang determines a model as an embedding model through its `hf_config.architectures`. But gte model has the same architecture as CausalLM. So I added a new parameter in the `server_args` and changed the forward function of `Qwen2ForCausalLM`.

## Modifications

1. Changed the forward function of `Qwen2ForCausalLM`.
2. Added a new parameter `is_embedding` in `server_args`.
3. Some related changes.
4. Added unit tests for gte models. (both in the generation and embedding tests. I used rouge-L score in the generation tests)
5. Changed readme.

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -222,6 +229,8 @@ python -m sglang.launch_server --model-path meta-llama/Meta-Llama-3-8B-Instruct
  
 ### Supported Models
 
+**Completion Models**
```
Comment (Ying1123) [inline README.md]:
```suggestion
**Generative Models**
```

File: README.md [hunk 2]
```diff
@@ -242,6 +251,12 @@ python -m sglang.launch_server --model-path meta-llama/Meta-Llama-3-8B-Instruct
 - ChatGLM
 - InternLM 2
 
+**Embedding Models**
+
+- e5-mistral
+- gte-Qwen2
+  - `python -m sglang.launch_server --model-path intfloat/e5-mistral-7b-instruct --is-embedding`
```
Comment (Ying1123) [inline README.md]:
```suggestion
  - `python -m sglang.launch_server --model-path Alibaba-NLP/gte-Qwen2-7B-instruct --is-embedding`
```

File: python/sglang/srt/server_args.py [hunk 3]
```diff
@@ -201,6 +202,12 @@ def add_cli_args(parser: argparse.ArgumentParser):
             action="store_true",
             help="Whether or not to allow for custom models defined on the Hub in their own modeling files.",
         )
+        parser.add_argument(
+            "--is-embedding",
+            action="store_true",
+            default=False,
```
Comment (Ying1123) [inline python/sglang/srt/server_args.py]:
```suggestion
```

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 4]
```diff
@@ -94,7 +94,11 @@ def __init__(
             trust_remote_code=server_args.trust_remote_code,
             model_overide_args=model_overide_args,
         )
-        self.is_generation = is_generation_model(self.hf_config.architectures)
+
+        assert self.server_args.is_embedding is not None
```
Comment (Ying1123) [inline python/sglang/srt/managers/tokenizer_manager.py]:
```suggestion
```

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 5]
```diff
@@ -94,7 +94,11 @@ def __init__(
...
+        self.is_generation = is_generation_model(
+            self.hf_config.architectures, self.server_args
```
Comment (Ying1123) [inline python/sglang/srt/managers/tokenizer_manager.py]:
```suggestion
            self.hf_config.architectures, self.server_args.is_embedding
```

File: python/sglang/srt/utils.py [hunk 6]
```diff
@@ -224,12 +224,17 @@ def is_multimodal_model(model):
     raise ValueError("unrecognized type")
 
 
-def is_generation_model(model_architectures):
+def is_generation_model(model_architectures, server_args=None):
```
Comment (Ying1123) [inline python/sglang/srt/utils.py]:
```suggestion
def is_generation_model(model_architectures, is_embedding=False):
```

File: python/sglang/srt/utils.py [hunk 7]
```diff
@@ -224,12 +224,17 @@ def is_multimodal_model(model):
...
+    # We have two ways to determine whether a model is a generation model.
+    # 1. Check if the architectue
```
Comment (Ying1123) [inline python/sglang/srt/utils.py]:
```suggestion
    # 1. Check the model architectue
```

File: python/sglang/srt/utils.py [hunk 8]
```diff
@@ -224,12 +224,17 @@ def is_multimodal_model(model):
     raise ValueError("unrecognized type")
 
 
-def is_generation_model(model_architectures):
+def is_generation_model(model_architectures, server_args=None):
+    # We have two ways to determine whether a model is a generation model.
```
Comment (Ying1123) [inline python/sglang/srt/utils.py]:
```suggestion
    # We have two ways to determine whether a model is a generative model.
```

File: python/sglang/srt/utils.py [hunk 9]
```diff
@@ -224,12 +224,17 @@ def is_multimodal_model(model):
...
+    # 1. Check if the architectue
+    # 2. check the server args
```
Comment (Ying1123) [inline python/sglang/srt/utils.py]:
```suggestion
    # 2. check the `is_embedding` in server args
```

File: python/sglang/srt/utils.py [hunk 10]
```diff
@@ -224,12 +224,17 @@ def is_multimodal_model(model):
...
     if (
         "LlamaEmbeddingModel" in model_architectures
         or "MistralModel" in model_architectures
     ):
         return False
+    if server_args is not None:
+        return not server_args.is_embedding
```
Comment (Ying1123) [inline python/sglang/srt/utils.py]:
```suggestion
    return not server_args.is_embedding
```

File: python/sglang/srt/utils.py [hunk 11]
```diff
@@ -224,12 +224,17 @@ def is_multimodal_model(model):
...
     return True
```
Comment (Ying1123) [inline python/sglang/srt/utils.py]:
```suggestion
```

File: python/sglang/srt/managers/tp_worker.py [hunk 12]
```diff
@@ -844,6 +845,7 @@ def run_tp_server(
 ):
     """Run a tensor parallel server."""
     try:
+
```
Comment (Ying1123) [inline python/sglang/srt/managers/tp_worker.py]:
```suggestion
```

File: python/sglang/srt/managers/tp_worker.py [hunk 13]
```diff
@@ -871,6 +873,7 @@ def launch_tp_servers(
     """Launch multiple tensor parallel servers."""
     procs = []
     for i in tp_rank_range:
+
```
Comment (Ying1123) [inline python/sglang/srt/managers/tp_worker.py]:
```suggestion
```

File: python/sglang/srt/model_executor/model_runner.py [hunk 14]
```diff
@@ -196,7 +196,7 @@ def load_model(self):
             else None
         )
         self.is_generation = is_generation_model(
-            self.model_config.hf_config.architectures
+            self.model_config.hf_config.architectures, self.server_args
```
Comment (Ying1123) [inline python/sglang/srt/model_executor/model_runner.py]:
```suggestion
            self.model_config.hf_config.architectures, self.server_args.is_embedding
```

File: python/sglang/srt/model_executor/model_runner.py [hunk 15]
```diff
@@ -498,9 +498,18 @@ def forward_extend(self, batch: ScheduleBatch):
             batch,
             forward_mode=ForwardMode.EXTEND,
         )
-        return self.model.forward(
-            batch.input_ids, input_metadata.positions, input_metadata
-        )
+        if self.is_generation == True:
```
Comment (Ying1123) [inline python/sglang/srt/model_executor/model_runner.py]:
```suggestion
        if self.is_generation:
```

File: python/sglang/srt/models/qwen2.py [hunk 16]
```diff
@@ -283,11 +285,15 @@ def forward(
...
         hidden_states = self.model(input_ids, positions, input_metadata, input_embeds)
-        return self.logits_processor(
-            input_ids, hidden_states, self.lm_head.weight, input_metadata
-        )
+        if get_embedding:
+            return self.pooler(hidden_states, input_metadata)
+        else:
+            return self.logits_processor(
+                input_ids, hidden_states, self.lm_head.weight, input_metadata
+            )
```
Comment (Ying1123) [inline python/sglang/srt/models/qwen2.py:294]: Change the order of two branches. We always choose the generation branch as the first one.

File: python/sglang/srt/server.py [hunk 17]
```diff
@@ -307,6 +307,7 @@ def launch_server(
                 (server_args.node_rank + 1) * tp_size_local,
             )
         )
+        #! 这里 launch 了 tp server
```
Comment (Ying1123) [inline python/sglang/srt/server.py]:
```suggestion
```

File: python/sglang/srt/server.py [hunk 18]
```diff
@@ -500,6 +503,7 @@ def __init__(
     ):
         """See the arguments in server_args.py::ServerArgs"""
         self.server_args = ServerArgs(*args, log_level=log_level, **kwargs)
+        assert self.server_args.is_embedding is not None, "is_embedding must be set."
```
Comment (Ying1123) [inline python/sglang/srt/server.py]:
```suggestion
```

File: test/srt/models/test_generation_models.py
Comment (Ying1123) [inline test/srt/models/test_generation_models.py:1]: Add `Alibaba-NLP/gte-Qwen2-1.5B-instruct` to this test.

**Review by Ying1123 — CHANGES_REQUESTED** [review]

**Comment by zhaochenyang20** [comment]
@Ying1123 I added gte in the generation model test. Note that I changed the prefill tolerance accordingly and added the rouge-l metric instead of assert output_strs exactly the same.

**Review by Ying1123 — APPROVED** [review]

**Review by Ying1123 — APPROVED** [review]

File: test/srt/models/test_generation_models.py [hunk 19]
```diff
@@ -56,17 +94,34 @@ def assert_close_prefill_logits_and_output_strs(
...
+        assert all(
+            score >= rouge_threshold for score in rouge_l_scores
+        ), f"Not all ROUGE-L scores are greater than {rouge_threshold}"
 
     def test_prefill_logits_and_output_strs(self):
-        for model, tp_size, long_context_tolerance in MODELS:
+        import multiprocessing as mp
+
+        try:
+            mp.set_start_method("spawn")
```
Comment (hnyls2002) [inline test/srt/models/test_generation_models.py:114]: Why would this be needed?

**Comment by llmforever** [comment]
@zhaochenyang20
您好，使用您提交的这个方法，和原始transformer与sentence transformer得到的embedding差距都很大，并且效果不好，您能帮忙看看吗？7B/1.5B的模型都测试过了 
different result compare with orginal transformer backend,why?

prompt = “hello world”
sglang:
import openai
client = openai.Client(
    base_url="http://localhost:30000/v1", api_key="EMPTY")

response = client.embeddings.create(
    model="default",
    input=prompt ,
)

transformer：
tokenizer = AutoTokenizer.from_pretrained('Alibaba-NLP/gte-Qwen2-7B-instruct', trust_remote_code=True)
model = AutoModel.from_pretrained('Alibaba-NLP/gte-Qwen2-7B-instruct', trust_remote_code=True)

max_length = 8192

batch_dict = tokenizer(prompt, max_length=max_length, padding=True, truncation=True, return_tensors='pt')
outputs = model(**batch_dict)
embeddings = last_token_pool(outputs.last_hidden_state, batch_dict['attention_mask'])

embeddings = F.normalize(embeddings, p=2, dim=1)

**Comment by zhaochenyang20** [comment]
@llmforever hello. Sorry, I haven't noticed this before. Do you still need to fix this? Actually, we have a unit test for this in `test/srt/models/test_embedding_models.py`. The logits here is indeed closed.

Also, I don't understand what did you mean by "perform not so well". Could you provide your running snifts and your serving command for SGLang.

And, does e5-mistral also have this problem? Or only get?

**Comment by thomZ1** [comment]
image
The same problem, I tried using the SGLang OpenAI API and SentenceTransformer with the same prompt, but the output embeddings were different.

**Comment by zhaochenyang20** [comment]
Yeah. The embedding could be different due to a lot of reasons. @llmforever 

You can check this unit test: https://github.com/sgl-project/sglang/blob/main/test/srt/models/test_embedding_models.py

We set a tolerance value for the embedding difference.

Also, please try the e5-mistral model and give us the embedding difference. 

https://huggingface.co/intfloat/e5-mistral-7b-instruct

@Ying1123 Do you think the difference provided is tolerable?

**Comment by llmforever** [comment]
> Yeah. The embedding could be different due to a lot of reasons. @llmforever
> 
> You can check this unit test: https://github.com/sgl-project/sglang/blob/main/test/srt/models/test_embedding_models.py
> 
> We set a tolerance value for the embedding difference.
> 
> Also, please try the e5-mistral model and give us the embedding difference.
> 
> https://huggingface.co/intfloat/e5-mistral-7b-instruct
> 
> @Ying1123 Do you think the difference provided is tolerable?

I test about 10 cases，**each accuracy drop from 80% to less than 10%**，i think the difference is not tolerable，but the result of the e5-mistral-7b-instruct model is the same，can you please help me look that? Here is the code i use to generate the embedding：

**for transformer：**

import torch
import torch.nn.functional as F

from torch import Tensor
from transformers import AutoTokenizer, AutoModel

input_texts = ['hello']
tokenizer = AutoTokenizer.from_pretrained('Alibaba-NLP/gte-Qwen2-7B-instruct', trust_remote_code=True)
model = AutoModel.from_pretrained('Alibaba-NLP/gte-Qwen2-7B-instruct', trust_remote_code=True)

max_length = 8192

batch_dict = tokenizer(input_texts, max_length=max_length, padding=True, truncation=True, return_tensors='pt')
outputs = model(**batch_dict)
embeddings = last_token_pool(outputs.last_hidden_state, batch_dict['attention_mask'])

embeddings = F.normalize(embeddings, p=2, dim=1)

**for sglang:**
import openai
import time
client = openai.Client(
    base_url="http://localhost:30000/v1", api_key="EMPTY")

input_texts = ['hello']

queres = client.embeddings.create(
    model="default",
    input=quelist,
)
embeddings  = torch.tensor(response.data[0].embedding)

**Comment by zhaochenyang20** [comment]
@Ying1123 I think he provides an intolerable difference hummm?  I gonna check it these days.
